File: lobster-input-ios/LobsterKeyboard/Pinyin/PinyinFuzzy.swift
import Foundation

/// 模糊音 + 容错纠错(引擎层无状态工具)。与安卓 PinyinFuzzy.kt 对齐。
/// 模糊音:把合法音节派生出等价音节集合(平翘舌/前后鼻音/边鼻音…),按开关启用。
/// 纠错:对单个音节做编辑距离≤1(替换/插入/删除/邻键)还原成合法音节,仅在精确/模糊不足时兜底。
enum PinyinFuzzy {

    struct Settings {
        var zZh = false; var cCh = false; var sSh = false
        var lN = false; var fH = false; var rL = false
        var anAng = false; var enEng = false; var inIng = false
        var correction = false

        var anyInitial: Bool { zZh || cCh || sSh || lN || fH || rL }
        var anyFinal: Bool { anAng || enEng || inIng }
        var anyFuzzy: Bool { anyInitial || anyFinal }
    }

    private static let neighbors: [Character: String] = [
        "q": "wa", "w": "qes", "e": "wrd", "r": "etf", "t": "ryg",
        "y": "tuh", "u": "yij", "i": "uok", "o": "ipl", "p": "ol",
        "a": "qsz", "s": "awdzx", "d": "serfcx", "f": "drtgvc",
        "g": "ftyhbv", "h": "gyujnb", "j": "huikmn", "k": "jiolm",
        "l": "kop", "z": "asx", "x": "zsdc", "c": "xdfv",
        "v": "cfgb", "b": "vghn", "n": "bhjm", "m": "njk"
    ]

    /// 把一个合法音节派生出模糊等价音节集合(含自身)。调用方需对模糊命中降权。
    static func variants(_ syllable: String, _ s: Settings) -> [String] {
        if !s.anyFuzzy || syllable.isEmpty { return [syllable] }
        var set: [String] = [syllable]
        var seen: Set<String> = [syllable]
        func add(_ v: String) { if seen.insert(v).inserted { set.append(v) } }

        if s.anyInitial {
            for v in set {
                if s.zZh { if v.hasPrefix("zh") { add("z" + v.dropFirst(2)) } else if v.hasPrefix("z") { add("zh" + v.dropFirst(1)) } }
                if s.cCh { if v.hasPrefix("ch") { add("c" + v.dropFirst(2)) } else if v.hasPrefix("c") { add("ch" + v.dropFirst(1)) } }
                if s.sSh { if v.hasPrefix("sh") { add("s" + v.dropFirst(2)) } else if v.hasPrefix("s") { add("sh" + v.dropFirst(1)) } }
                if s.lN { if v.hasPrefix("l") { add("n" + v.dropFirst(1)) } else if v.hasPrefix("n") { add("l" + v.dropFirst(1)) } }
                if s.fH { if v.hasPrefix("f") { add("h" + v.dropFirst(1)) } else if v.hasPrefix("h") { add("f" + v.dropFirst(1)) } }
                if s.rL { if v.hasPrefix("r") { add("l" + v.dropFirst(1)) } else if v.hasPrefix("l") { add("r" + v.dropFirst(1)) } }
            }
        }
        if s.anyFinal {
            for v in set {
                if s.anAng { if v.hasSuffix("ang") { add(String(v.dropLast(3)) + "an") } else if v.hasSuffix("an") { add(String(v.dropLast(2)) + "ang") } }
                if s.enEng { if v.hasSuffix("eng") { add(String(v.dropLast(3)) + "en") } else if v.hasSuffix("en") { add(String(v.dropLast(2)) + "eng") } }
                if s.inIng { if v.hasSuffix("ing") { add(String(v.dropLast(3)) + "in") } else if v.hasSuffix("in") { add(String(v.dropLast(2)) + "ing") } }
            }
        }
        return set
    }

    /// 对一个"非法音节"做编辑距离≤1 纠错,返回可能的合法音节(由 isSyllable 校验)。结果上限 8。
    static func corrections(_ token: String, _ isSyllable: (String) -> Bool) -> [String] {
        let chars = Array(token)
        if chars.count < 2 || chars.count > 6 { return [] }
        var out: [String] = []
        var seen = Set<String>()
        func add(_ s: String) { if isSyllable(s) && seen.insert(s).inserted { out.append(s) } }
        // 删除一个字母
        for i in chars.indices { var c = chars; c.remove(at: i); add(String(c)) }
        // 替换为邻键
        for i in chars.indices {
            for n in (neighbors[chars[i]] ?? "") {
                var c = chars; c[i] = n; add(String(c))
            }
            if out.count >= 8 { return out }
        }
        // 插入一个字母
        let letters = Array("abcdefghijklmnopqrstuvwxyz")
        for i in 0...chars.count {
            for ch in letters {
                var c = chars; c.insert(ch, at: i); add(String(c))
                if out.count >= 8 { return out }
            }
        }
        return out
    }
}

File: lobster-input-android/app/src/main/java/com/lobster/input/keyboard/pinyin/PinyinFuzzy.kt
package com.lobster.input.keyboard.pinyin

/**
 * 模糊音 + 容错纠错(引擎层无状态工具)。
 *
 * 设计对齐 RIME/搜狗:模糊音与纠错分两层、各自可控。
 *  - 模糊音:把一个合法音节派生出"等价音节集合"(平翘舌、前后鼻音、边鼻音…),按开关启用。
 *  - 纠错:对单个音节做编辑距离≤1(替换/插入/删除/邻键)还原成合法音节,仅在精确/模糊不足时兜底。
 *
 * 注意:模糊音派生只对"合法音节"做,声母/韵母规则覆盖 i/u 介音变体
 *  (an/ang 规则天然覆盖 ian/iang、uan/uang;故无需单列)。
 */
object PinyinFuzzy {

    /** 模糊音开关集合(默认全关,由用户在设置页逐项开启)。 */
    data class Settings(
        val zZh: Boolean = false,   // z ↔ zh
        val cCh: Boolean = false,   // c ↔ ch
        val sSh: Boolean = false,   // s ↔ sh
        val lN: Boolean = false,    // l ↔ n
        val fH: Boolean = false,    // f ↔ h
        val rL: Boolean = false,    // r ↔ l
        val anAng: Boolean = false, // an ↔ ang (含 ian/iang、uan/uang)
        val enEng: Boolean = false, // en ↔ eng
        val inIng: Boolean = false, // in ↔ ing
        val correction: Boolean = false // 容错纠错(编辑距离≤1)
    ) {
        val anyInitial get() = zZh || cCh || sSh || lN || fH || rL
        val anyFinal get() = anAng || enEng || inIng
        val anyFuzzy get() = anyInitial || anyFinal
    }

    /** 邻键表(QWERTY),用于纠错的邻键替换。 */
    private val neighbors = mapOf(
        'q' to "wa", 'w' to "qes", 'e' to "wrd", 'r' to "etf", 't' to "ryg",
        'y' to "tuh", 'u' to "yij", 'i' to "uok", 'o' to "ipl", 'p' to "ol",
        'a' to "qsz", 's' to "awdz x", 'd' to "serfcx", 'f' to "drtgvc",
        'g' to "ftyhbv", 'h' to "gyujnb", 'j' to "huikmn", 'k' to "jiolm",
        'l' to "kop", 'z' to "asx", 'x' to "zsdc", 'c' to "xdfv",
        'v' to "cfgb", 'b' to "vghn", 'n' to "bhjm", 'm' to "njk"
    )

    /**
     * 把一个合法音节派生出模糊等价音节集合(含自身)。
     * 仅用于在候选查词时扩展;调用方需对模糊命中做降权。
     */
    fun variants(syllable: String, s: Settings): Set<String> {
        if (!s.anyFuzzy || syllable.isEmpty()) return setOf(syllable)
        var set = linkedSetOf(syllable)
        // ---- 声母 ----
        if (s.anyInitial) {
            val add = LinkedHashSet<String>()
            for (v in set) {
                // 双向:对每个开启的对,长前缀优先判定
                if (s.zZh) { if (v.startsWith("zh")) add.add("z" + v.substring(2)) else if (v.startsWith("z")) add.add("zh" + v.substring(1)) }
                if (s.cCh) { if (v.startsWith("ch")) add.add("c" + v.substring(2)) else if (v.startsWith("c")) add.add("ch" + v.substring(1)) }
                if (s.sSh) { if (v.startsWith("sh")) add.add("s" + v.substring(2)) else if (v.startsWith("s")) add.add("sh" + v.substring(1)) }
                if (s.lN) { if (v.startsWith("l")) add.add("n" + v.substring(1)) else if (v.startsWith("n")) add.add("l" + v.substring(1)) }
                if (s.fH) { if (v.startsWith("f")) add.add("h" + v.substring(1)) else if (v.startsWith("h")) add.add("f" + v.substring(1)) }
                if (s.rL) { if (v.startsWith("r")) add.add("l" + v.substring(1)) else if (v.startsWith("l")) add.add("r" + v.substring(1)) }
            }
            set.addAll(add)
        }
        // ---- 韵母(前后鼻音) ----
        if (s.anyFinal) {
            val add = LinkedHashSet<String>()
            for (v in set) {
                if (s.anAng) { if (v.endsWith("ang")) add.add(v.dropLast(3) + "an") else if (v.endsWith("an")) add.add(v.dropLast(2) + "ang") }
                if (s.enEng) { if (v.endsWith("eng")) add.add(v.dropLast(3) + "en") else if (v.endsWith("en")) add.add(v.dropLast(2) + "eng") }
                if (s.inIng) { if (v.endsWith("ing")) add.add(v.dropLast(3) + "in") else if (v.endsWith("in")) add.add(v.dropLast(2) + "ing") }
            }
            set.addAll(add)
        }
        return set
    }

    /**
     * 对一个"非法音节"做编辑距离≤1 纠错,返回可能的合法音节(由 isSyllable 校验)。
     * 仅覆盖:邻键替换、漏字母(插入)、多字母(删除)。结果上限 8。
     */
    fun corrections(token: String, isSyllable: (String) -> Boolean): List<String> {
        if (token.length !in 2..6) return emptyList()
        val out = LinkedHashSet<String>()
        val letters = "abcdefghijklmnopqrstuvwxyz"
        // 删除一个字母
        for (i in token.indices) {
            val c = token.removeRange(i, i + 1)
            if (c.length >= 1 && isSyllable(c)) out.add(c)
        }
        // 替换为邻键
        for (i in token.indices) {
            for (n in neighbors[token[i]].orEmpty()) {
                if (n == ' ') continue
                val c = token.substring(0, i) + n + token.substring(i + 1)
                if (isSyllable(c)) out.add(c)
            }
            if (out.size >= 8) return out.toList()
        }
        // 插入一个字母
        for (i in 0..token.length) {
            for (ch in letters) {
                val c = token.substring(0, i) + ch + token.substring(i)
                if (isSyllable(c)) out.add(c)
                if (out.size >= 8) return out.toList()
            }
        }
        return out.toList()
    }
}
